### PR TITLE
feat(day2): add harbor configuration cr

### PR DESCRIPTION
**Pull request by chlins**
Closes https://github.com/goharbor/harbor-operator/issues/444

Signed-off-by: chlins <chlins.zhang@gmail.com>

### CODE REVIEW

**Comment by chlins** [comment]
```yaml
apiVersion: v1
data:
  email_password: YmFyCg==
kind: Secret
metadata:
  name: secret-sample
  namespace: cluster-sample-ns
type: Opaque
---

apiVersion: goharbor.io/v1beta1
kind: HarborConfiguration
metadata:
  labels:
    harbor-name: harborcluster-sample-harbor
  name: test-config
  namespace: cluster-sample-ns
  resourceVersion: "258368"
  selfLink: /apis/goharbor.io/v1beta1/namespaces/cluster-sample-ns/harborconfigurations/test-config
  uid: aa21c53f-31e8-4fff-999d-b8407c6ede5d
spec:
  email_password: secret-sample
  email_ssl: true
status:
  lastApplyTime: "2021-07-09T06:07:26Z"
  lastConfiguration:
    email_password: secret-sample
    email_ssl: true
  status: Success
```

File: apis/goharbor.io/v1beta1/harbor_configuration.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+	Spec HarborConfigurationSpec `json:"spec,omitempty"`
+
+	Status HarborConfigurationStatus `json:"status,omitempty"`
+}
+
+// HarborConfigurationSpec defines the spec of HarborConfiguration.
+type HarborConfigurationSpec struct {
+	// AdditionalProperties provides workaround for those unknown configuration fields in the future.
+	// +kubebuilder:validation:Optional
+	Extension `json:",inline"`
```
Comment (cndoit18) [inline apis/goharbor.io/v1beta1/harbor_configuration.go:79]: If the user fills in auth_mode, is it recorded to the AuthMode field or this field?

Comment (steven-zou) [reply apis/goharbor.io/v1beta1/harbor_configuration.go:79]: @chlins 

Any answers here?

**Comment by steven-zou** [comment]
@chlins 

lint failures and conflicts should be fixed ASAP.

File: apis/goharbor.io/v1beta1/harbor_configuration.go (hunk 1 above)
Comment (chlins) [reply apis/goharbor.io/v1beta1/harbor_configuration.go:79]: Extension will catch all field,  so maybe we cannot use the json inline format, extension should as external field. Code updated.

**Comment by bitsf** [comment]
lost `bases/goharbor.io_harborconfigurations.yaml` in `config/crd/kustomization.yaml`

**Review by bitsf — APPROVED** [review]

File: docs/configurations/day2-config.md [hunk 2]
```diff
@@ -1,14 +1,16 @@
 # Day2 configurations
 
-## Step 1
+Initially, we configured harbor by means of configmap, but currently we recommend using `harborconfiguration` crd to configure harbor, for the configmap method will be deprecated in version 2.3, those who have used and still use configmap will be automatically converted to `harborconfiguration` cr by the controller, and automatically remove old configmap.
```
Comment (steven-zou) [inline docs/configurations/day2-config.md:3]:
```suggestion
Initially, we configured harbor by means of configmap, but currently we recommend using `harborconfiguration` crd to configure harbor, for the configmap method will be deprecated in version 1.3, those who have used and still use configmap will be automatically converted to `harborconfiguration` cr by the controller, and automatically remove old configmap.
```

Comment (steven-zou) [reply docs/configurations/day2-config.md:3]: @chlins 1.3？

**Review by steven-zou — CHANGES_REQUESTED** [review]
do changes?

File: docs/configurations/day2-config.md (hunk 2 above)
Comment (chlins) [reply docs/configurations/day2-config.md]: yes, changed.

**Review by steven-zou — APPROVED** [review]
LGTM

**Pull request merged**